src/main/java/org/opensearch/knn/index/query/nativelib/DocAndScoreQuery.java

```diff
@@ -0,0 +1,185 @@
+                            }
+
+                            @Override
+                            public long cost() {
+                                return upper - lower;
+                            }
+                        };
+                    }
+
+                    @Override
+                    public float getMaxScore(int docId) {
+                        docId += context.docBase;
+                        float maxScore = 0;
+                        for (int idx = Math.max(0, upTo); idx < upper && docs[idx] <= docId; idx++) {
+                            maxScore = Math.max(maxScore, scores[idx]);
+                        }
+                        return maxScore * boost;
+                    }
+
+                    @Override
+                    public float score() {
+                        return scores[upTo] * boost;
+                    }
+
+                    @Override
+                    public int advanceShallow(int docid) {
+                        int start = Math.max(upTo, lower);
+                        int docidIndex = Arrays.binarySearch(docs, start, upper, docid + context.docBase);
+                        if (docidIndex < 0) {
+                            docidIndex = -1 - docidIndex;
+                        }
+                        if (docidIndex >= upper) {
+                            return NO_MORE_DOCS;
+                        }
+                        return docs[docidIndex];
+                    }
+
+                    /**
+                     * move the implementation of docID() into a differently-named method so we can call it
+                     * from DocIDSetIterator.docID() even though this class is anonymous
+                     *
+                     * @return the current docid
+                     */
+                    private int docIdNoShadow() {
+                        if (upTo == -1) {
+                            return -1;
+                        }
+                        if (upTo >= upper) {
+                            return NO_MORE_DOCS;
+                        }
+                        return docs[upTo] - context.docBase;
+                    }
+
+                    @Override
+                    public int docID() {
+                        return docIdNoShadow();
+                    }
+                };
+            }
+
+            @Override
+            public boolean isCacheable(LeafReaderContext ctx) {
+                return true;
+            }
+        };
+    }
+
+    @Override
+    public String toString(String field) {
+        return "DocAndScore[" + k + "][docs:" + Arrays.toString(docs) + ", scores:" + Arrays.toString(scores) + "]";
+    }
+
+    @Override
+    public void visit(QueryVisitor visitor) {
+        visitor.visitLeaf(this);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (!sameClassAs(obj)) {
+            return false;
+        }
+        return contextIdentity == ((DocAndScoreQuery) obj).contextIdentity
```

Comment:
nit:- can we just use equals method from contextIndentity

Reply:
saves us from npe. will keep it as it is